yo1k/tic_tac_toe/core/game.py

```diff
@@ -0,0 +1,296 @@
+from enum import Enum, auto
+from collections.abc import Sequence
+from typing import Optional
+from abc import ABC, abstractmethod
+
+
+class Mark(Enum):
+    X = auto()
+    O = auto()
+
+
+class Player:
+    def __init__(self, mark: Mark):
+        self.__mark = mark
+        self.__wins = 0
+
+    def mark(self) -> Mark:
+        return self.__mark
+
+    @property
+    def wins(self) -> int:
+        return self.__wins
+
+    @wins.setter
+    def wins(self, wins: int):
+        self.__wins = wins
+
+    def __str__(self):
+        return f"{type(self)}(" \
+               f"mark={self.__mark}," \
+               f"wins={self.__wins})"
+
+
+class Phase(Enum):
+    ONSET = auto()
+    INROUND = auto()
+    OUTROUND = auto()
+
+
+class Cell:
+    """A game board cell."""
+
+    def __init__(self, x: int, y: int):
+        assert 0 <= x < Board.size(), f"{x}"
+        assert 0 <= y < Board.size(), f"{y}"
+        self.__x = x
+        self.__y = y
+
+    @property
+    def x(self) -> int:
+        return self.__x
+
+    @property
+    def y(self) -> int:
+        return self.__y
+
+    def __str__(self):
+        return f"{type(self)}(" \
+               f"x={self.__x}," \
+               f"y={self.__y})"
+
+
+class Board:
+    def __init__(self, cells: Sequence[list[Optional[Mark]]] = None):
+        self.__cells = Board.__empty_cells() if cells is None else cells
+
+    def _set(self, cell: Cell, mark: Mark):
+        self.__cells[cell.x][cell.y] = mark
+
+    def _get(self, cell: Cell) -> Optional[Mark]:
+        return self.__cells[cell.x][cell.y]
+
+    def _clean(self):
+        self.__cells = Board.__empty_cells()
+
+    @property
+    def cells(self):
+        return self.__cells
+
+    @staticmethod
+    def size() -> int:
+        return 3
+
+    @staticmethod
+    def __empty_cells() -> list[list[Optional[Mark]]]:
+        return [[None for _ in range(Board.size())] for _ in range(Board.size())]
+
+    def __str__(self):
+        return f"{type(self)}(" \
+               f"cells={self.__cells},"
+
+
+class State:
+    """Full game state which is enough to restore a paused game."""
+
+    def __init__(self,
+                 game_rounds: int,
+                 players: Sequence[Player],
+                 board: Board,
+                 phase=Phase.ONSET,
+                 cur_round: int = 0,
+                 step: int = 0):
+        self.game_rounds = game_rounds
+        self.round = cur_round
+        self.players = players
```

Comment:
Let's add an assertion that the size of `players` is 2.

Reply:
Done.